src/main/scala/org/allenai/pipeline/Pipeline.scala

```diff
@@ -1,75 +1,206 @@
 
-  protected[this] val persistedSteps: ListBuffer[Producer[_]] = ListBuffer()
+  def rootOutputUrl: URI
 
   /** Run the pipeline.  All steps that have been persisted will be computed, along with any upstream dependencies */
   def run(title: String) = {
-    runPipelineReturnResults(title, persistedSteps.toSeq)
+    runPipelineReturnResults(title, persistedSteps)
   }
 
-  def runOne[T](target: Producer[T]) =
-    runOnly(target.stepInfo.className, List(target): _*).toList(0).asInstanceOf[T]
+  def persistedSteps = steps.toMap
+
+  /** Common-case methods for persisting Producers */
+  object Persist {
+
+    /** Persist a collection */
+    object Collection {
+      def asText[T: StringSerializable: ClassTag](
+        step: Producer[Iterable[T]],
+        stepName: String = null,
+        suffix: String = ".txt"
+      )(): PersistedProducer[Iterable[T], FlatArtifact] =
+        persist(step, LineCollectionIo.text[T], stepName, suffix)
+
+      def asJson[T: JsonFormat: ClassTag](
+        step: Producer[Iterable[T]],
+        stepName: String = null,
+        suffix: String = ".json"
+      )(): PersistedProducer[Iterable[T], FlatArtifact] =
+        persist(step, LineCollectionIo.json[T], stepName, suffix)
+    }
+
+    /** Persist a single object */
+    object Singleton {
+      def asText[T: StringSerializable: ClassTag](
+        step: Producer[T],
+        stepName: String = null,
+        suffix: String = ".txt"
+      )(): PersistedProducer[T, FlatArtifact] =
+        persist(step, SingletonIo.text[T], stepName, suffix)
+
+      def asJson[T: JsonFormat: ClassTag](
+        step: Producer[T],
+        stepName: String = null,
+        suffix: String = ".json"
+      )(): PersistedProducer[T, FlatArtifact] =
+        persist(step, SingletonIo.json[T], stepName, suffix)
+    }
+
+    /** Persist an Iterator */
+    object Iterator {
+      def asText[T: StringSerializable: ClassTag](
+        step: Producer[Iterator[T]],
+        stepName: String = null,
+        suffix: String = ".txt"
+      ): PersistedProducer[Iterator[T], FlatArtifact] =
+        persist(step, LineIteratorIo.text[T], stepName, suffix)
+
+      def asJson[T: JsonFormat: ClassTag](
+        step: Producer[Iterator[T]],
+        stepName: String = null,
+        suffix: String = ".json"
+      )(): PersistedProducer[Iterator[T], FlatArtifact] =
+        persist(step, LineIteratorIo.json[T], stepName, suffix)
+    }
+
+  }
+
+  /** Create a persisted version of the given Producer
+    * The producer is registered as a target of the pipeline, and will be computed
+    * when the pipeline is run.
+    * See Persist.Collection, Persist.Singleton, etc. utility methods above
+    * @param original the non-persisted Producer
+    * @param io the serialization format
+    * @param suffix a file suffix
+    * @return the persisted Producer
+    */
+  def persist[T, A <: Artifact: ClassTag](
```

Comment:
Pipeline.persist methods now all have a name argument. This is used to name the output Artifact and can be used to retrieve the step by name using Pipeline.persistedSteps